// @splitsoftware/splitio/es/__tests__/nodeSuites/impressions-listener.spec.js
import _Object$defineProperty from "@babel/runtime-corejs3/core-js-stable/object/define-property";
import _Object$defineProperties from "@babel/runtime-corejs3/core-js-stable/object/define-properties";
import _Object$getOwnPropertyDescriptors from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors";
import _forEachInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/for-each";
import _Object$getOwnPropertyDescriptor from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor";
import _filterInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/filter";
import _Object$getOwnPropertySymbols from "@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols";
import _Object$keys from "@babel/runtime-corejs3/core-js-stable/object/keys";
import _defineProperty from "@babel/runtime-corejs3/helpers/esm/defineProperty";
import _setTimeout from "@babel/runtime-corejs3/core-js-stable/set-timeout";

function ownKeys(object, enumerableOnly) { var keys = _Object$keys(object); if (_Object$getOwnPropertySymbols) { var symbols = _Object$getOwnPropertySymbols(object); if (enumerableOnly) symbols = _filterInstanceProperty(symbols).call(symbols, function (sym) { return _Object$getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { var _context; _forEachInstanceProperty(_context = ownKeys(Object(source), true)).call(_context, function (key) { _defineProperty(target, key, source[key]); }); } else if (_Object$getOwnPropertyDescriptors) { _Object$defineProperties(target, _Object$getOwnPropertyDescriptors(source)); } else { var _context2; _forEachInstanceProperty(_context2 = ownKeys(Object(source))).call(_context2, function (key) { _Object$defineProperty(target, key, _Object$getOwnPropertyDescriptor(source, key)); }); } } return target; }

import sinon from 'sinon';
import { SplitFactory } from '../../';
import SettingsFactory from '../../utils/settings';
var settings = SettingsFactory({
  core: {
    key: '<fake id>'
  }
});
var listener = {
  logImpression: sinon.stub()
};
var config = {
  core: {
    authorizationKey: '<fake-token-3>'
  },
  scheduler: {
    featuresRefreshRate: 1,
    segmentsRefreshRate: 1,
    metricsRefreshRate: 3000,
    impressionsRefreshRate: 3000
  },
  startup: {
    eventsFirstPushWindow: 3000
  },
  impressionListener: listener
};
export default function (assert) {
  var splitio = SplitFactory(config);
  var client = splitio.client();
  return client.ready().then(function () {
    var metaData = {
      ip: settings.runtime.ip,
      hostname: settings.runtime.hostname,
      sdkLanguageVersion: settings.version
    };
    var testAttrs = {
      is_test: true
    }; // Generate one impression, depends on hierarchical_dep_hierarchical which depends on hierarchical_dep_always_on

    client.getTreatment('nicolas@split.io', 'hierarchical_splits_test');
    client.getTreatment({
      matchingKey: 'marcio@split.io',
      bucketingKey: 'impr_bucketing_2'
    }, 'qc_team');
    client.getTreatment('facundo@split.io', 'qc_team', testAttrs);

    _setTimeout(function () {
      assert["true"](listener.logImpression.calledThrice, 'Impression listener logImpression method should be called after we call client.getTreatment, once per each impression generated.');
      assert["true"](listener.logImpression.getCall(0).calledWithMatch(_objectSpread({
        impression: {
          feature: 'hierarchical_splits_test',
          keyName: 'nicolas@split.io',
          treatment: 'on',
          bucketingKey: undefined,
          label: 'expected label'
        },
        attributes: undefined
      }, metaData)));
      assert["true"](listener.logImpression.getCall(1).calledWithMatch(_objectSpread({
        impression: {
          feature: 'qc_team',
          keyName: 'marcio@split.io',
          treatment: 'no',
          bucketingKey: 'impr_bucketing_2',
          label: 'default rule'
        },
        attributes: undefined
      }, metaData)));
      assert["true"](listener.logImpression.getCall(2).calledWithMatch(_objectSpread({
        impression: {
          feature: 'qc_team',
          keyName: 'facundo@split.io',
          treatment: 'no',
          bucketingKey: undefined,
          label: 'default rule'
        },
        attributes: testAttrs
      }, metaData)));
      client.destroy();
      assert.end();
    }, 0);
  });
}
