// @splitsoftware/splitio/lib/storage/ImpressionsCache/InRedis.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

require("core-js/modules/es.array.iterator");

require("core-js/modules/es.object.to-string");

require("core-js/modules/web.dom-collections.iterator");

require("core-js/modules/web.url.to-json");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _stringify = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/json/stringify"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _keys = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/keys"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/createClass"));

/**
Copyright 2016 Split Software

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/
var ImpressionsCacheInRedis =
/*#__PURE__*/
function () {
  function ImpressionsCacheInRedis(keys, redis, meta) {
    (0, _classCallCheck2.default)(this, ImpressionsCacheInRedis);
    this.keys = keys;
    this.redis = redis;
    this.meta = meta;
  }

  (0, _createClass2.default)(ImpressionsCacheInRedis, [{
    key: "track",
    value: function track(impressions) {
      var _this = this;

      return this.redis.rpush((0, _keys.default)(this).buildImpressionsKey(), this.toJSON(impressions)).then(function (queuedCount) {
        // If this is the creation of the key on Redis, set the expiration for it in 1hr.
        if (queuedCount === impressions.length) {
          return _this.redis.expire((0, _keys.default)(_this).buildImpressionsKey(), 3600);
        }
      });
    }
  }, {
    key: "toJSON",
    value: function toJSON(impressions) {
      var _this2 = this;

      return (0, _map.default)(impressions).call(impressions, function (impression) {
        var keyName = impression.keyName,
            bucketingKey = impression.bucketingKey,
            feature = impression.feature,
            treatment = impression.treatment,
            label = impression.label,
            time = impression.time,
            changeNumber = impression.changeNumber;
        return (0, _stringify.default)({
          m: _this2.meta,
          i: {
            k: keyName,
            b: bucketingKey,
            f: feature,
            t: treatment,
            r: label,
            c: changeNumber,
            m: time
          }
        });
      });
    }
    /**
     * We are returning true because the go syncronizer push the impressions from redis.
     */

  }, {
    key: "isEmpty",
    value: function isEmpty() {
      return true;
    }
  }]);
  return ImpressionsCacheInRedis;
}();

var _default = ImpressionsCacheInRedis;
exports.default = _default;
