// @splitsoftware/splitio/lib/utils/context/constants.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = exports.STORAGE = exports.STATUS_MANAGER = exports.SETTINGS = exports.READY_FROM_CACHE = exports.READY = exports.READINESS = exports.EVENTS = exports.DESTROYED = exports.COLLECTORS = void 0;
var COLLECTORS = 'metrics_collectors';
exports.COLLECTORS = COLLECTORS;
var DESTROYED = 'is_destroyed';
exports.DESTROYED = DESTROYED;
var EVENTS = 'events_publisher';
exports.EVENTS = EVENTS;
var READINESS = 'readiness_gate';
exports.READINESS = READINESS;
var READY = 'is_ready';
exports.READY = READY;
var READY_FROM_CACHE = 'is_ready_from_cache';
exports.READY_FROM_CACHE = READY_FROM_CACHE;
var SETTINGS = 'settings';
exports.SETTINGS = SETTINGS;
var STATUS_MANAGER = 'status_manager';
exports.STATUS_MANAGER = STATUS_MANAGER;
var STORAGE = 'storage';
exports.STORAGE = STORAGE;
var _default = {
  COLLECTORS: COLLECTORS,
  DESTROYED: DESTROYED,
  EVENTS: EVENTS,
  READINESS: READINESS,
  READY: READY,
  READY_FROM_CACHE: READY_FROM_CACHE,
  SETTINGS: SETTINGS,
  STATUS_MANAGER: STATUS_MANAGER,
  STORAGE: STORAGE
};
exports.default = _default;
