// @splitsoftware/splitio/lib/__tests__/nodeSuites/impressions.spec.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

require("core-js/modules/es.regexp.constructor");

require("core-js/modules/es.regexp.exec");

require("core-js/modules/es.regexp.to-string");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = _default;

var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs3/regenerator"));

var _now = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/date/now"));

var _some = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/some"));

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/filter"));

require("regenerator-runtime/runtime");

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/asyncToGenerator"));

var _ = require("../../");

var _settings = _interopRequireDefault(require("../../utils/settings"));

var _labels = require("../../utils/labels");

var _splitchangesSince = _interopRequireDefault(require("../mocks/splitchanges.since.-1.json"));

var _splitchangesSince2 = _interopRequireDefault(require("../mocks/splitchanges.since.1457552620999.json"));

var baseUrls = {
  sdk: 'https://sdk.baseurl/impressionsSuite',
  events: 'https://events.baseurl/impressionsSuite'
};
var settings = (0, _settings.default)({
  core: {
    key: '<fake id>'
  },
  urls: baseUrls
});
var config = {
  core: {
    authorizationKey: '<fake-token-2>'
  },
  scheduler: {
    featuresRefreshRate: 1,
    segmentsRefreshRate: 1,
    metricsRefreshRate: 3000,
    impressionsRefreshRate: 5
  },
  urls: baseUrls,
  startup: {
    eventsFirstPushWindow: 3000
  }
};

function _default(_x, _x2, _x3) {
  return _ref.apply(this, arguments);
}

function _ref() {
  _ref = (0, _asyncToGenerator2.default)(
  /*#__PURE__*/
  _regenerator.default.mark(function _callee(key, mock, assert) {
    var splitio, client, evaluationsStart, readyEvaluationsStart, evaluationsEnd;
    return _regenerator.default.wrap(function _callee$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            // Mocking this specific route to make sure we only get the items we want to test from the handlers.
            mock.onGet(settings.url('/splitChanges?since=-1')).replyOnce(200, _splitchangesSince.default);
            mock.onGet(settings.url('/splitChanges?since=1457552620999')).reply(200, _splitchangesSince2.default);
            mock.onGet(new RegExp("".concat(settings.url('/segmentChanges/'), ".*"))).reply(200, {
              since: 10,
              till: 10,
              name: 'segmentName',
              added: [],
              removed: []
            });
            splitio = (0, _.SplitFactory)(config);
            client = splitio.client();
            evaluationsStart = 0, readyEvaluationsStart = 0, evaluationsEnd = 0;
            mock.onPost(settings.url('/testImpressions/bulk')).replyOnce(function (req) {
              var _context;

              var data = JSON.parse(req.data);
              assert.equal(data.length, 3, 'We performed evaluations for three splits, so we should have 3 items total.'); // finding these validate the feature names collection too

              var dependencyChildImpr = (0, _filter.default)(data).call(data, function (e) {
                return e.testName === 'hierarchical_splits_test';
              })[0];
              var alwaysOnWithConfigImpr = (0, _filter.default)(data).call(data, function (e) {
                return e.testName === 'split_with_config';
              })[0];
              var notExistentSplitImpr = (0, _filter.default)(data).call(data, function (e) {
                return e.testName === 'not_existent_split';
              })[0];
              assert.equal(notExistentSplitImpr.keyImpressions.length, 1); // Only one, the split not found is filtered by the non existent Split check.

              assert.equal(alwaysOnWithConfigImpr.keyImpressions.length, 1);
              assert.equal(dependencyChildImpr.keyImpressions.length, 1);
              assert.true(dependencyChildImpr, 'Split we wanted to evaluate should be present on the impressions.');
              assert.false((0, _some.default)(data).call(data, function (e) {
                return e.testName === 'hierarchical_dep_always_on';
              }), 'Parent split evaluations should not result in impressions.');
              assert.false((0, _some.default)(data).call(data, function (e) {
                return e.testName === 'hierarchical_dep_hierarchical';
              }), 'No matter how deep is the chain.');
              assert.true(alwaysOnWithConfigImpr, 'Split evaluated with config should have generated an impression too.');
              assert.false(Object.prototype.hasOwnProperty.call(alwaysOnWithConfigImpr.keyImpressions[0], 'configuration'), 'Impressions do not change with configuration evaluations.');
              assert.false(Object.prototype.hasOwnProperty.call(alwaysOnWithConfigImpr.keyImpressions[0], 'config'), 'Impressions do not change with configuration evaluations.');

              function validateImpressionData(output, expected) {
                var performedWhenReady = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
                assert.equal(output.keyName, expected.keyName, 'Present impressions should have the correct key.');
                assert.equal(output.bucketingKey, expected.bucketingKey, 'Present impressions should have the correct bucketingKey.');
                assert.equal(output.treatment, expected.treatment, 'Present impressions should have the correct treatment.');
                assert.equal(output.label, expected.label, 'Present impressions should have the correct label.');
                assert.equal(output.changeNumber, expected.changeNumber, 'Present impressions should have the correct changeNumber.');
                assert.true(output.time >= (performedWhenReady ? readyEvaluationsStart : evaluationsStart) && output.time <= evaluationsEnd, 'Present impressions should have the correct timestamp (test with error margin).');
              }

              validateImpressionData(notExistentSplitImpr.keyImpressions[0], {
                keyName: 'facundo@split.io',
                label: _labels.SDK_NOT_READY,
                treatment: 'control',
                bucketingKey: undefined,
                changeNumber: undefined
              }, false);
              validateImpressionData(dependencyChildImpr.keyImpressions[0], {
                keyName: 'facundo@split.io',
                label: 'expected label',
                treatment: 'on',
                bucketingKey: undefined,
                changeNumber: 2828282828
              });
              validateImpressionData(alwaysOnWithConfigImpr.keyImpressions[0], {
                keyName: 'facundo@split.io',
                label: 'another expected label',
                treatment: 'on',
                bucketingKey: 'test_buck_key',
                changeNumber: 828282828282
              }); // Not push impressions with a invalid key (aka matching key)

              assert.true((0, _filter.default)(_context = dependencyChildImpr.keyImpressions).call(_context, function (e) {
                return e.keyName !== 'facundo@split.io';
              }).length === 0, 'There should be impressions with valid keys, the sdk will not push a impression with invalid a key');
              client.destroy();
              assert.end();
              return [200];
            });
            splitio.Logger.enable();
            evaluationsStart = (0, _now.default)();
            assert.equal(client.getTreatment(key, 'not_existent_split'), 'control', "If we try to get an evaluation BEFORE the client is ready, we expect ".concat(_labels.SDK_NOT_READY, " label on the impression."));
            _context2.next = 12;
            return client.ready();

          case 12:
            readyEvaluationsStart = (0, _now.default)(); // depends on hierarchical_dep_hierarchical which depends on hierarchical_dep_always_on

            assert.equal(client.getTreatment(key, 'not_existent_split'), 'control', 'If we try to get an evaluation for a non existent split AFTER the client is ready, it won\'t log an impression.');
            assert.equal(client.getTreatment(key, 'hierarchical_splits_test'), 'on', 'We should get an evaluation as always.');
            assert.equal(client.getTreatment(false, 'hierarchical_splits_test'), 'control', 'We should return control with a invalid key and that impression is not tracked.');
            assert.deepEqual(client.getTreatmentWithConfig({
              matchingKey: key,
              bucketingKey: 'test_buck_key'
            }, 'split_with_config'), {
              // I'll run this one with bucketing key.
              treatment: 'on',
              config: '{"color":"brown","dimensions":{"height":12,"width":14},"text":{"inner":"click me"}}'
            }, 'We should get an evaluation as always.');
            splitio.Logger.disable();
            evaluationsEnd = (0, _now.default)();

          case 19:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee);
  }));
  return _ref.apply(this, arguments);
}
