// @splitsoftware/splitio/es/__tests__/browserSuites/ignore-ip-addresses-setting.spec.js
import _forEachInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/for-each";
import _regeneratorRuntime from "@babel/runtime-corejs3/regenerator";
import { SplitFactory } from '../..';
import SettingsFactory from '../../utils/settings'; // Header keys and expected values. Expected values are obtained with the runtime function evaluated with IPAddressesEnabled in true.

var HEADER_SPLITSDKMACHINEIP = 'SplitSDKMachineIP';
var HEADER_SPLITSDKMACHINENAME = 'SplitSDKMachineName'; // Refresh rates are set to 1 second to finish the test quickly. Otherwise, it would finish in 1 minute (60 seconds is the default value)

var scheduler = {
  metricsRefreshRate: 1,
  impressionsRefreshRate: 1,
  eventsPushRate: 1
}; // Config with IPAddressesEnabled set to false

var configWithIPAddressesDisabled = {
  core: {
    authorizationKey: '<fake-token>',
    key: 'nicolas@split.io',
    IPAddressesEnabled: false
  },
  urls: {
    sdk: 'https://sdk.split-ipdisabled.io/api',
    events: 'https://events.split-ipdisabled.io/api'
  },
  scheduler: scheduler
}; // Config with IPAddressesEnabled set to true

var configWithIPAddressesEnabled = {
  core: {
    authorizationKey: '<fake-token>',
    key: 'nicolas@split.io',
    IPAddressesEnabled: true
  },
  urls: {
    sdk: 'https://sdk.split-ipenabled.io/api',
    events: 'https://events.split-ipenabled.io/api'
  },
  scheduler: scheduler
}; // Config with default IPAddressesEnabled (true)

var configWithIPAddressesDefault = {
  core: {
    authorizationKey: '<fake-token>',
    key: 'nicolas@split.io'
  },
  urls: {
    sdk: 'https://sdk.split-ipdefault.io/api',
    events: 'https://events.split-ipdefault.io/api'
  },
  scheduler: scheduler
};
var configSamples = [configWithIPAddressesDisabled, configWithIPAddressesEnabled, configWithIPAddressesDefault];
var postEndpoints = ['/events/bulk', '/testImpressions/bulk', '/metrics/times', '/metrics/counters'];
export default function (mock, assert) {
  // Generator to synchronize the call of assert.end() when all Splitio configurations are run.
  var finish =
  /*#__PURE__*/
  _regeneratorRuntime.mark(function _callee() {
    var CONFIG_SAMPLES_COUNT, i;
    return _regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            CONFIG_SAMPLES_COUNT = configSamples.length;
            i = 0;

          case 2:
            if (!(i < CONFIG_SAMPLES_COUNT - 1)) {
              _context.next = 8;
              break;
            }

            _context.next = 5;
            return;

          case 5:
            i++;
            _context.next = 2;
            break;

          case 8:
            assert.end();

          case 9:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  })(); // Assert request headers


  function assertHeaders(IPAddressesEnabled, req) {
    assert["false"](HEADER_SPLITSDKMACHINEIP in req.headers, "Request must not include ".concat(HEADER_SPLITSDKMACHINEIP, " header, no matters the value of IPAddressesEnabled."));
    assert["false"](HEADER_SPLITSDKMACHINENAME in req.headers, "Request must not include ".concat(HEADER_SPLITSDKMACHINENAME, " header, no matters the value of IPAddressesEnabled."));
  }

  function mockAndAssertIPAddressesEnabled(config) {
    // Assert properties in impressions logged to impression listener
    config.impressionListener = {
      logImpression: function logImpression(impression) {
        assert["false"](impression.ip, '"ip" property in impressions must be false, no matters the value of IPAddressesEnabled.');
        assert["false"](impression.hostname, '"hostname" property in impressions must be false, no matters the value of IPAddressesEnabled.');
      }
    }; // Init Split client

    var splitio = SplitFactory(config);
    var client = splitio.client();
    var settings = SettingsFactory(config); // Generator to synchronize the destruction of the client when all the post endpoints where called once.

    var finishConfig =
    /*#__PURE__*/
    _regeneratorRuntime.mark(function _callee2() {
      var POST_ENDPOINTS_TO_TEST, i;
      return _regeneratorRuntime.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              POST_ENDPOINTS_TO_TEST = postEndpoints.length;
              i = 0;

            case 2:
              if (!(i < POST_ENDPOINTS_TO_TEST - 1)) {
                _context2.next = 8;
                break;
              }

              _context2.next = 5;
              return;

            case 5:
              i++;
              _context2.next = 2;
              break;

            case 8:
              client.destroy();
              finish.next();

            case 10:
            case "end":
              return _context2.stop();
          }
        }
      }, _callee2);
    })(); // Mock and assert POST endpoints


    _forEachInstanceProperty(postEndpoints).call(postEndpoints, function (postEndpoint) {
      mock.onPost(settings.url(postEndpoint)).replyOnce(function (req) {
        assertHeaders(settings.core.IPAddressesEnabled, req);
        finishConfig.next();
        return [200];
      });
    }); // Run normal client flow 


    client.ready().then(function () {
      client.getTreatment('hierarchical_splits_test');
      client.track('sometraffictype', 'someEvent', 10);
    });
  }

  _forEachInstanceProperty(configSamples).call(configSamples, function (configSample) {
    return mockAndAssertIPAddressesEnabled(configSample);
  });
}
