// @splitsoftware/splitio/lib/utils/logger/LoggerFactory.js
/* eslint-disable no-console */
'use strict';

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.Logger = exports.setLogLevel = exports.LogLevels = void 0;

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _keys = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/keys"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _assign = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/assign"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/createClass"));

var isNode = typeof process !== 'undefined' && process.version ? true : false;
var LogLevels = {
  'DEBUG': 'DEBUG',
  'INFO': 'INFO',
  'WARN': 'WARN',
  'ERROR': 'ERROR',
  'NONE': 'NONE'
}; // DEBUG is the default. The log level is not specific to an SDK instance.

exports.LogLevels = LogLevels;
var GlobalLogLevel = LogLevels.DEBUG;

var setLogLevel = function setLogLevel(level) {
  GlobalLogLevel = level;
};

exports.setLogLevel = setLogLevel;
var defaultOptions = {
  showLevel: true,
  displayAllErrors: false
};

var Logger =
/*#__PURE__*/
function () {
  function Logger(category, options) {
    (0, _classCallCheck2.default)(this, Logger);
    this.category = category;
    this.options = (0, _assign.default)({}, defaultOptions, options);
  }

  (0, _createClass2.default)(Logger, [{
    key: "debug",
    value: function debug(msg) {
      if (this._shouldLog(LogLevels.DEBUG)) this._log(LogLevels.DEBUG, msg);
    }
  }, {
    key: "info",
    value: function info(msg) {
      if (this._shouldLog(LogLevels.INFO)) this._log(LogLevels.INFO, msg);
    }
  }, {
    key: "warn",
    value: function warn(msg) {
      if (this._shouldLog(LogLevels.WARN)) this._log(LogLevels.WARN, msg);
    }
  }, {
    key: "error",
    value: function error(msg) {
      if (this.options.displayAllErrors || this._shouldLog(LogLevels.ERROR)) this._log(LogLevels.ERROR, msg);
    }
  }, {
    key: "_log",
    value: function _log(level, text) {
      var formattedText = this._generateLogMessage(level, text);

      var method = level === LogLevels.ERROR && !isNode ? 'error' : 'log';
      console[method](formattedText);
    }
  }, {
    key: "_generateLogMessage",
    value: function _generateLogMessage(level, text) {
      var textPre = ' => ';
      var result = '';

      if (this.options.showLevel) {
        result += '[' + level + ']' + (level === LogLevels.INFO || level === LogLevels.WARN ? ' ' : '') + ' ';
      }

      if (this.category) {
        result += this.category + textPre;
      }

      return result += text;
    }
  }, {
    key: "_shouldLog",
    value: function _shouldLog(level) {
      var _context;

      var logLevel = GlobalLogLevel;
      var levels = (0, _map.default)(_context = (0, _keys.default)(LogLevels)).call(_context, function (f) {
        return LogLevels[f];
      });
      var index = (0, _indexOf.default)(levels).call(levels, level); // What's the index of what it's trying to check if it should log

      var levelIdx = (0, _indexOf.default)(levels).call(levels, logLevel); // What's the current log level index.

      return index >= levelIdx;
    }
  }]);
  return Logger;
}();

exports.Logger = Logger;
