// @splitsoftware/splitio/lib/utils/constants/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.NA = exports.UNKNOWN = exports.CONTROL_WITH_CONFIG = exports.CONTROL = exports.STORAGE_LOCALSTORAGE = exports.STORAGE_REDIS = exports.STORAGE_MEMORY = exports.CONSUMER_MODE = exports.PRODUCER_MODE = exports.STANDALONE_MODE = exports.LOCALHOST_MODE = void 0;
// SDK Modes
var LOCALHOST_MODE = 'localhost';
exports.LOCALHOST_MODE = LOCALHOST_MODE;
var STANDALONE_MODE = 'standalone';
exports.STANDALONE_MODE = STANDALONE_MODE;
var PRODUCER_MODE = 'producer';
exports.PRODUCER_MODE = PRODUCER_MODE;
var CONSUMER_MODE = 'consumer'; // Storage types

exports.CONSUMER_MODE = CONSUMER_MODE;
var STORAGE_MEMORY = 'MEMORY';
exports.STORAGE_MEMORY = STORAGE_MEMORY;
var STORAGE_REDIS = 'REDIS';
exports.STORAGE_REDIS = STORAGE_REDIS;
var STORAGE_LOCALSTORAGE = 'LOCALSTORAGE'; // Special treatments

exports.STORAGE_LOCALSTORAGE = STORAGE_LOCALSTORAGE;
var CONTROL = 'control';
exports.CONTROL = CONTROL;
var CONTROL_WITH_CONFIG = {
  treatment: CONTROL,
  config: null
}; // Constants for unknown and not-applicable values

exports.CONTROL_WITH_CONFIG = CONTROL_WITH_CONFIG;
var UNKNOWN = 'unknown';
exports.UNKNOWN = UNKNOWN;
var NA = 'NA';
exports.NA = NA;
