// @splitsoftware/splitio/lib/trackers/impressions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _setTimeout2 = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/set-timeout"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _splice = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/splice"));

var _logger = _interopRequireDefault(require("../utils/logger"));

var _thenable = _interopRequireDefault(require("../utils/promise/thenable"));

/**
Copyright 2016 Split Software

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/
var log = (0, _logger.default)('splitio-client:impressions-tracker');

function ImpressionsTrackerContext(context) {
  var collector = context.get(context.constants.STORAGE).impressions;
  var settings = context.get(context.constants.SETTINGS);
  var listener = settings.impressionListener;
  var _settings$runtime = settings.runtime,
      ip = _settings$runtime.ip,
      hostname = _settings$runtime.hostname;
  var sdkLanguageVersion = settings.version;
  var _queue = [];
  return {
    queue: function queue(impression, attributes) {
      _queue.push({
        impression: impression,
        attributes: attributes
      });
    },
    track: function track() {
      var impressionsCount = _queue.length;
      var slice = (0, _splice.default)(_queue).call(_queue, 0, impressionsCount);
      var res = collector.track((0, _map.default)(slice).call(slice, function (_ref) {
        var impression = _ref.impression;
        return impression;
      })); // If we're on an async storage, handle error and log it.

      if ((0, _thenable.default)(res)) {
        res.then(function () {
          var _context;

          log.debug((0, _concat.default)(_context = "Successfully stored ".concat(impressionsCount, " impression")).call(_context, impressionsCount === 1 ? '' : 's', "."));
        }).catch(function (err) {
          var _context2, _context3;

          log.error((0, _concat.default)(_context2 = (0, _concat.default)(_context3 = "Could not store impressions bulk with ".concat(impressionsCount, " impression")).call(_context3, impressionsCount === 1 ? '' : 's', ". Error: ")).call(_context2, err));
        });
      } // Wrap in a timeout because we don't want it to be blocking.


      var _loop = function _loop(i) {
        listener && (0, _setTimeout2.default)(function () {
          try {
            // An exception on the listener should not break the SDK.
            listener.logImpression({
              impression: slice[i].impression,
              attributes: slice[i].attributes,
              ip: ip,
              hostname: hostname,
              sdkLanguageVersion: sdkLanguageVersion
            });
          } catch (err) {
            log.error("Impression listener logImpression method threw: ".concat(err, "."));
          }
        }, 0);
      };

      for (var i = 0; i < impressionsCount; i++) {
        _loop(i);
      }
    }
  };
}

var _default = ImpressionsTrackerContext;
exports.default = _default;
