// @splitsoftware/splitio/es/storage/ImpressionsCache/InRedis.js
import _JSON$stringify from "@babel/runtime-corejs3/core-js-stable/json/stringify";
import _mapInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/map";
import _keysInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/keys";
import _classCallCheck from "@babel/runtime-corejs3/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime-corejs3/helpers/esm/createClass";

/**
Copyright 2016 Split Software

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/
var ImpressionsCacheInRedis =
/*#__PURE__*/
function () {
  function ImpressionsCacheInRedis(keys, redis, meta) {
    _classCallCheck(this, ImpressionsCacheInRedis);

    this.keys = keys;
    this.redis = redis;
    this.meta = meta;
  }

  _createClass(ImpressionsCacheInRedis, [{
    key: "track",
    value: function track(impressions) {
      var _this = this;

      return this.redis.rpush(_keysInstanceProperty(this).buildImpressionsKey(), this.toJSON(impressions)).then(function (queuedCount) {
        // If this is the creation of the key on Redis, set the expiration for it in 1hr.
        if (queuedCount === impressions.length) {
          return _this.redis.expire(_keysInstanceProperty(_this).buildImpressionsKey(), 3600);
        }
      });
    }
  }, {
    key: "toJSON",
    value: function toJSON(impressions) {
      var _this2 = this;

      return _mapInstanceProperty(impressions).call(impressions, function (impression) {
        var keyName = impression.keyName,
            bucketingKey = impression.bucketingKey,
            feature = impression.feature,
            treatment = impression.treatment,
            label = impression.label,
            time = impression.time,
            changeNumber = impression.changeNumber;
        return _JSON$stringify({
          m: _this2.meta,
          i: {
            k: keyName,
            b: bucketingKey,
            f: feature,
            t: treatment,
            r: label,
            c: changeNumber,
            m: time
          }
        });
      });
    }
    /**
     * We are returning true because the go syncronizer push the impressions from redis.
     */

  }, {
    key: "isEmpty",
    value: function isEmpty() {
      return true;
    }
  }]);

  return ImpressionsCacheInRedis;
}();

export default ImpressionsCacheInRedis;
