// @splitsoftware/splitio-react/lib/index.js
"use strict";
exports.__esModule = true;
// Split SDK factory (Renamed to avoid name conflict with SplitFactory component)
var splitio_1 = require("@splitsoftware/splitio");
exports.SplitSdk = splitio_1.SplitFactory;
// HOC functions
var withSplitFactory_1 = require("./withSplitFactory");
exports.withSplitFactory = withSplitFactory_1["default"];
var withSplitClient_1 = require("./withSplitClient");
exports.withSplitClient = withSplitClient_1["default"];
var withSplitTreatments_1 = require("./withSplitTreatments");
exports.withSplitTreatments = withSplitTreatments_1["default"];
// Render props components
var SplitTreatments_1 = require("./SplitTreatments");
exports.SplitTreatments = SplitTreatments_1["default"];
var SplitClient_1 = require("./SplitClient");
exports.SplitClient = SplitClient_1["default"];
var SplitFactory_1 = require("./SplitFactory");
exports.SplitFactory = SplitFactory_1["default"];
// helper functions/hooks
var useClient_1 = require("./useClient");
exports.useClient = useClient_1["default"];
var useTreatments_1 = require("./useTreatments");
exports.useTreatments = useTreatments_1["default"];
var useTrack_1 = require("./useTrack");
exports.useTrack = useTrack_1["default"];
var useManager_1 = require("./useManager");
exports.useManager = useManager_1["default"];
// SplitContext
var SplitContext_1 = require("./SplitContext");
exports.SplitContext = SplitContext_1["default"];
